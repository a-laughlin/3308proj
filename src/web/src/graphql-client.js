import React,{createElement} from 'react';
import ApolloClient, { gql } from "apollo-boost";
import {useObservable} from './hooks';
import {pipe,cond,renameProps,ife,isString} from './utils';
// import {envs} from '../../../envs';
/**
 * the Apollo client handles data caching+exchange with the Apollo graphql server
 */
export {gql};

export const uri = process.env.NODE_ENV==='production'?
  '/graphql' :
  'http://localhost:4000/graphql';

export const client = new ApolloClient({uri});

// client.defaultOptions

/** fetchPolicy https://github.com/apollographql/apollo-client/blob/master/packages/apollo-client/src/core/watchQueryOptions.ts#L10
 * fetchPolicy determines where the client may return a result from. The options are:
 * - cache-first (default): return result from cache. Only fetch from network if cached result is not available.
 * - cache-and-network: return result from cache first (if it exists), then return network result once it's available.
 * - cache-only: return result from cache if available, fail otherwise.
 * - no-cache: return result from network, fail if network call doesn't succeed, don't save to cache
 * - network-only: return result from network, fail if network call doesn't succeed, save to cache
 * - standby: only for queries that aren't actively watched, but should be available for refetch and updateQueries.
 */

/** errorPolicy https://github.com/apollographql/apollo-client/blob/master/packages/apollo-client/src/core/watchQueryOptions.ts#L27
 * errorPolicy determines the level of events for errors in the execution result. The options are:
 * - none (default): any errors from the request are treated like runtime errors and the observable is stopped (XXX this is default to lower breaking changes going from AC 1.0 => 2.0)
 * - ignore: errors from the request do not stop the observable, but also don't call `next`
 * - all: errors are treated like data and will notify observables
 */

client.defaultOptions = {
  // errorPolicy: ignore (no errors)|none(error or data)|all (errors and data)
  watchQuery: {fetchPolicy:'cache-first',errorPolicy:'none'},
  query: {fetchPolicy:'cache-first',errorPolicy:'none'},
  mutate: {fetchPolicy:'no-cache',errorPolicy:'none'}
};

/* Initial State */
client.writeData({data:{}})

/** networkStatus from query responses
 * https://github.com/apollographql/apollo-client/blob/master/packages/apollo-client/src/core/networkStatus.ts
 */
// options from https://www.apollographql.com/docs/react/api/apollo-client.html#ApolloClient.watchQuery
export const getUseWatchQuery = (options={})=>{
  const obs = client.watchQuery(options);
  return function useWatchQuery(props){return useObservable(obs,{data:{},loading:true});}
}

export const isLoading = x=>x.loading;
export const isError = x=>x.errors && x.errors.length;
export const isData = x=>!isLoading(x) && !isError(x);

export const useHeartRateQuery = variables=>getUseWatchQuery({
  query:gql`query ($summary_date: ID!,$steps:Int,$model_id: ID){
    heartRatePredictions (summary_date: $summary_date,steps:$steps,model_id:$model_id){
      summary_date
      bedtime_start
      rates
      history{
        rates
      }
    }
  }`,
  variables,
});
