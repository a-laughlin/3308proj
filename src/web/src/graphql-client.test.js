import React from 'react';
import ReactDOM from 'react-dom';
// import {App,SvgMain} from './App';
import {act} from 'react-testing-library';
import { useState } from 'react';
import { renderHook } from 'react-hooks-testing-library';
import { useHeartRateQuery,uri,client,gql } from './graphql-client';
// describe('useStateExample', () => {
//   it('uses state', () => {
//     // const {result,unmount,waitForNextUpdate} = renderHook(() => useState('foo'));
//     const {result,waitForNextUpdate,unmount} = renderHook(() => useState('foo'));
//     expect(result.current[0]).toBe('foo')
//     act(() => result.current[1]('bar'))
//     expect(result.current[0]).toBe('bar')
//     // obj.unmount(); // only necessary with useffect
//   })
// });

// current (any) - the return value of the callback function
// error (Error) - the error that was thrown if the callback function threw an error during rendering
// waitForNextUpdate (function) - returns a Promise that resolves the next time the hook renders, commonly when state is updated as the result of a asynchronous action.
// rerender (function([newProps])) - function to rerender the test component including any hooks called in the callback function. If newProps are passed, the will replace the initialProps passed the the callback function for future renders.
// unmount (function()) - function to unmount the test component, commonly used to trigger cleanup effects for useEffect hooks.
describe('useHeartRateQuery hook', () => {
  it('renders correct data', async () => {

    const {waitForNextUpdate,result,unmount} = renderHook(()=>
      useHeartRateQuery({summary_date:"2018-11-06",steps:20})({}));
    expect(result.current.loading).toBe(true);
    await act(waitForNextUpdate)
    expect(result.current.loading).toBe(false);
    expect(result.current.data).toMatchSnapshot();
    unmount();
  });
  // need error test now that tests are working
});
