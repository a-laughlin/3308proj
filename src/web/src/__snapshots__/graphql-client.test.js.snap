// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useHeartRateQuery hook renders correct data 1`] = `
Object {
  "heartRatePredictions": Array [
    Object {
      "__typename": "HeartRateList",
      "bedtime_start": "1970-01-01T00:00:30.012Z",
      "history": Object {
        "__typename": "HeartRateList",
        "rates": Array [
          75,
          76,
          74,
          75,
          75,
          76,
          75,
          75,
          74,
          73,
          73,
          72,
          74,
          72,
          64,
          64,
          64,
          66,
          66,
          67,
          68,
          63,
          63,
          66,
          64,
          63,
          65,
          62,
          61,
          61,
          63,
          62,
          63,
          64,
          61,
          61,
          61,
          60,
          60,
          60,
          63,
          58,
          58,
          59,
          60,
          61,
          61,
          62,
          62,
          57,
          58,
          58,
          57,
          58,
          62,
          59,
          58,
          57,
          58,
          59,
          60,
          61,
          62,
          61,
          62,
          63,
          59,
          57,
          55,
          58,
          57,
          58,
          57,
          60,
          0,
          57,
          55,
          57,
          57,
          58,
          61,
          60,
          0,
          0,
          0,
          0,
          0,
          60,
          60,
          0,
          0,
        ],
      },
      "rates": Array [
        47,
        33,
        37,
        35,
        35,
        36,
        35,
        35,
        35,
        35,
        35,
        35,
        35,
        36,
        36,
        36,
        36,
        36,
        36,
        36,
        36,
      ],
      "summary_date": "2018-11-06_ml_model_foo",
    },
  ],
}
`;
